File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/NewsService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.news.NewsInquiryDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.ApplicationUser;
import at.ac.tuwien.sepm.groupphase.backend.entity.News;
import org.springframework.data.domain.Page;
import org.springframework.http.ResponseEntity;

public interface NewsService {

    /**
     * Publish a news entry.
     *
     * @param news to publish
     * @return published news entry
     */
    News publishNews(NewsInquiryDto news);

    /**
     * Find 20 news entries ordered by published at date (descending) on the page specified by {@code pageIndex}.
     *
     * @param pageIndex       index of page to load
     * @param loadAlreadyRead determines if news entries that have already been read by the user (true) or ones the user
     *                        has neven seen (false) should get loaded
     * @param user            specifies the user who requested to load the news page
     * @return page of 20 news entries ordered descending by the date of creation specified by {@code pageIndex}
     */
    Page<News> findAllPagedByCreatedAt(int pageIndex, boolean loadAlreadyRead, ApplicationUser user);

    /**
     * Fetch news entry with the specified id from the database.
     *
     * @param id of news to fetch
     * @return entity of news entry
     */
    News getById(Long id);

    /**
     * Delete news entry with the specified id and all many-to-many relations "newsRead" between a user and this news
     * from the database.
     *
     * @param id of news to delete
     */
    void deleteById(Long id);

    /**
     * Set relation "newsRead" between the {@code user} and the news with the id {@code id} if it hasn't already existed.
     *
     * @param id   of the news
     * @param user that has read the news
     */
    ResponseEntity<Void> putRelation(Long id, ApplicationUser user);
}

File: frontend/src/app/services/news/news.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Globals} from '../../global/globals';
import {AbbreviatedNews, News} from '../../dtos/news';


@Injectable({
  providedIn: 'root'
})
export class NewsService {

  private newsBaseUri: string = this.globals.backendUri + '/news';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }


  /**
   * Persists news with images to the backend
   *
   * @param news to persist
   * @return an Observable for the created news
   */
  create(news: News): Observable<News> {
    return this.httpClient.post<News>(
      this.newsBaseUri,
      news,
    );
  }


  /**
   * Loads a news-page from the backend
   *
   * @param pageIndex index of the page that should be fetched
   * @param loadAlreadyRead determines if news should get loaded that the user has already read or news that he has not read before
   * @return an Observable for the fetched page of news entries
   */
  getPage(pageIndex: number, loadAlreadyRead: boolean): Observable<AbbreviatedNews[]> {
    let params: HttpParams = new HttpParams();
    params = params.set('pageIndex', pageIndex);
    params = params.set('loadAlreadyRead', loadAlreadyRead);
    return this.httpClient.get<AbbreviatedNews[]>(this.newsBaseUri, {params});
  }


  /**
   * Fetch the news with the id "id" stored in the system and add relation news_read between user and news entry
   *
   * @param id The id of the news to fetch
   * @return observable of found news.
   */
  getById(id: number): Observable<News> {
    return this.httpClient.get<News>(this.newsBaseUri + `/${id}`);
  }

  /**
   * Delete the news with the id "id" stored in the system
   *
   * @param id The id of the news to delete
   * @return Not Found Exception if news with id "id" was not found
   */
  deleteById(id: number): Observable<any> {
    return this.httpClient.delete<News>(this.newsBaseUri + `/${id}`);
  }

  /**
   * Create a relation "news_read" between the user and the news with the id "id" if it hasn't existed already
   *
   * @param id The id of the news that was read
   * @return Not Found Exception if news with id "id" was not found
   */
  putNewsRead(id: number): Observable<any> {
    return this.httpClient.put<News>(this.newsBaseUri + `/${id}`, null);
  }
}
